core/drivers/crypto/crypto_api/acipher/rsa.c

```diff
@@ -485,6 +521,11 @@ TEE_Result crypto_acipher_rsassa_verify(uint32_t algo,
 		if (ret == TEE_ERROR_NOT_IMPLEMENTED)
 			ret = drvcrypt_rsassa_verify(&rsa_ssa);
 
+		if (ret == TEE_ERROR_NOT_IMPLEMENTED)
+			ret = sw_crypto_acipher_rsassa_verify(algo, key,
```

Comment:
Having this `sw_crypto_acipher_rsassa_verify()` software implementation as a fallback makes sense. But what I don't like so much is how the function is called. Here we're first trying to call one function to see if that works and if not we try another function. It's not that bad, it's just making matters slightly worse, especially where the hw-ops don't require any sw fallback.

I would have preferred that we called the right function from the start. Either by adding some extra parameters to `drvcrypt_get_ops()` so it can return an sw-ops pointer when needed or by letting the hw-ops functions themselves call `sw_crypto_acipher_rsassa_verify()` directly when needed. I'd prefer the first option.

Reply:
@jenswi-linaro I'll come back to you on this next week as soon as I am back in the office. thanks for your comments.

Reply:
IMO extending the the crypto API to support the fallback simplifies the overall project maintenance by unifying the behavior while also avoiding **code duplication in the platform specific drivers.**

Platforms not wanting to fallback to their software implementations for unsupported algorithms could perhaps return **TEE_ERROR_NOT_SUPPORTED** instead?

But I'd like to understand your concern a bit better though. Could you explain the issue as you see it please? is it a performance problem you are worried about?

Reply:
@jenswi-linaro also notice that the fallback might also occur on the key size not just the algorithm - so your first preference will not be applicable.

Reply:
The extra call is going to be required in any case:s ince it is a runtime decision for the key length and we cant associate behavior to a key size at some point a check is going to be needed. I mean we could but I dont think it is a good idea to get into that mess....it is going to make the code path unnecessarily complex.
The level of ugliness, well I see none on this case: IMO it is clean and easy to maintain. If I had made it obscure enough we probably wouldn't be having this argument.
And with respect to clarity, the hardware ops just need to return TEE_ERROR_NOT_IMPLEMENTED to fallback. And TEE_ERROR_NOT_SUPPORTED (proposal) to indicate that it wants to actually fail.

Seems clear enough to me.

Reply:
I'm not going to NACK this, but you'll need to find someone else to ack it.

Reply:
sure. I suppose @cneveux or @clementfaure might chip in with other alternatives? I dont want to be difficult but in this case is just that I truly disagree with you.

Reply:
Hi @clementfaure  and @jforissier as `drvcrypt_get_ops()` doesn't yet have the extra parameter in place at this time it seems that @ldts's patch is a clean and simple way to know that if the HW isn't implemented then it will fallback to the SW driver instead. In other words, it seems like adding any extra parameters to `drvcrypt_get_ops()` is a higher level OP-TEE issue that may be cleaner to address in a separate patch.

Reply:
Just one question @ldts about this implementation and one use case,
What happens if 
- The keypair is allocated and generated by the crypto HW driver (because the key size `size_bits` is supported).
- The crypto HW driver returns `TEE_ERROR_NOT_IMPLEMENTED` for the operations encrypt/decrypt and sign/verify (because the algo is not supported by the Crypto HW driver)
- The operation is redirected to the SW driver with a key allocated and generated by the Hardware driver.

How does this implementation work with a HW Crypto driver, where the private component of the allocated keypair is wrapped ? Typically, in the SE05X where the private component is a key_id and not a bignum key?